page-fault-handler/main.c

```diff
@@ -0,0 +1,910 @@
+#include <stdint.h>
+#include <stdbool.h>
+
+#include <hal/xbox.h>
+#include <xboxkrnl/xboxkrnl.h>
+#include <xboxrt/debug.h>
+#include <pbkit/pbkit.h>
+#include <hal/video.h>
+
+
+
+void payload_main(void);
+
+
+
+typedef uint32_t PageDirectoryEntry;
+typedef uint32_t PageTableEntry;
+
+// These are for both, PDEs and PTEs
+#define PDE_PRESENT_BIT (1 << 0)
+#define PDE_SIZE_BIT (1 << 7)
+
+//FIXME: From MSDN, should probably be in some of our headers?
+typedef struct _PCI_COMMON_CONFIG {
+  USHORT VendorID;
+  USHORT DeviceID;
+  USHORT Command; // 4
+  USHORT Status;
+  UCHAR  RevisionID; // 8
+  UCHAR  ProgIf; 
+  UCHAR  SubClass; // 10
+  UCHAR  BaseClass;
+  UCHAR  CacheLineSize; // 12
+  UCHAR  LatencyTimer;
+  UCHAR  HeaderType; // 14
+  UCHAR  BIST;
+  union {
+    struct {
+      ULONG BaseAddresses[6]; // 16, 20, 24, 28, 32, 36
+      ULONG Reserved1[2]; // 40, 44
+      ULONG ROMBaseAddress; // 48
+      ULONG Reserved2[2]; // 52, 56
+      UCHAR InterruptLine; // 60
+      UCHAR InterruptPin; 
+      UCHAR MinimumGrant; // 62
+      UCHAR MaximumLatency;
+    } type0;
+  } u;
+  UCHAR  DeviceSpecific[192]; // 64
+} __attribute__((packed)) PCI_COMMON_CONFIG;
+
+typedef struct {
+  uint16_t offset_lo;
+  uint16_t selector;
+  uint8_t zero;
+  uint8_t type_attr;
+  uint16_t offset_hi;
+} __attribute__((packed)) IDTEntry;
+
+typedef struct {
+  union {
+    uint32_t edi;
+    uint16_t di;
+  };
+  union {
+    uint32_t esi;
+    uint16_t si;
+  };
+  uint32_t ebp;
+  uint32_t _esp;
+  union {
+    uint32_t ebx;
+    struct {
+      union {
+        uint16_t bx;
+        struct {
+          uint8_t bl, bh;
+        };
+      };
+    };
+  };
+  union {
+    uint32_t edx;
+    struct {
+      union {
+        uint16_t dx;
+        struct {
+          uint8_t dl, dh;
+        };
+      };
+    };
+  };
+  union {
+    uint32_t ecx;
+    struct {
+      union {
+        uint16_t cx;
+        struct {
+          uint8_t cl, ch;
+        };
+      };
+    };
+  };
+  union {
+    uint32_t eax;
+    struct {
+      union {
+        uint16_t ax;
+        struct {
+          uint8_t al, ah;
+        };
+      };
+    };
+  };
+} __attribute__((packed)) Registers;
+
+typedef struct {
+  uint32_t error_code;
+  uint32_t eip;
+  uint16_t cs;
+  uint16_t _pad;
+  uint32_t eflags;
+} __attribute__((packed)) TrapFrame;
+
+
+uint32_t __stdcall do_test16(uint16_t a, uint16_t b, uint32_t eflags);
+asm("_do_test16@12:\n"
```

Comment:
The following code should be illegal!
The fact that we have 2 variations of the same function doesn't help. Someone should refactor this.